### ISSUE TITLE
mqtt_auth does not work

### ISSUE THREAD

**Issue by e-lin** [knolleary/pubsubclient#67]
I am working on an Arduino UNO project which is required an authorized password based on OAuth 1.0 to connects to the cloud via MQTT. This is alike [Authorizing a request to Twitter API][1], and [Creating a signature][2]. The whole process of creating a signature, here I said, password, requires algorithms like encodeURL, base64encode, and hmac-sha1. After processing, there is a base64-encoded, HMAC-SHA1 signature for the password.

However, I filled(even tried hard code) the generated password and username to the function client.connect("arduinoClient", "username", "password") in the example mqtt_auth, the Arduino device cannot connect to the cloud.

I've tried mqtt_basic example and the device connects to test.mosquitto.org very well. I've also tried to write a JAVA program that using my username and generated password to connect to the cloud via MQTT, and the JAVA program connects to the cloud successfully. I am pretty sure the username and password that I filled into the example mqtt_auth are right.

The MQTT broker of the cloud is implemented based on RabbitMQ.
I am not sure if there is an obvious mistake I've made here?

[1] https://dev.twitter.com/oauth/overview/authorizing-requests
[2] https://dev.twitter.com/oauth/overview/creating-signatures

**Comment by knolleary** [comment]
Hi,

can you share the code you used to set the password? It could well be a formatting/type issue.

**Comment by e-lin** [comment]
Hi @knolleary,

This is the part of JAVA code that for testing, and the password generated by this JAVA program can connect to the cloud.
```java
            Mac mac = Mac.getInstance(HMAC_SHA1_ALGORITHM);
            mac.init(keySpec);

            StringBuffer basestring = new StringBuffer();
            basestring.append(MQTT_BASE);
            basestring.append("&");
            basestring.append(URLEncoder.encode(URL_MAGELLAN_CRETENTIAL, "UTF-8"));
            basestring.append("&");
            basestring.append(URLEncoder.encode(auth.getParameterString(), "UTF-8"));
            byte[] result = mac.doFinal(basestring.toString().getBytes());
      
            BASE64Encoder encoder = new BASE64Encoder();
            String signature = encoder.encode(result);   
            auth.put(OAuthParameter.SIGNATURE, URLEncoder.encode(signature, "UTF-8"));
            String password = HEADER + ": " + auth.getHeaderValue();       
            password = encoder.encode(password.getBytes());
           
            MqttClient client = new MqttClient(apiUrl, "ArduinoSensor");
            MqttConnectOptions connOpts = new MqttConnectOptions();
            connOpts.setCleanSession(true);
            connOpts.setUserName(mCredential.getKey());
            connOpts.setPassword(password.toCharArray());
            //connOpts.setPassword("HARD-CODE-THE-GENERATED-PASSWORD".toCharArray());
            
            client.setCallback(new MqttCallback() {
                                  ....
            });
            client.connect(connOpts);
```
In this JAVA program, I can hard-code the generated password and connect to the cloud successfully.
But if I use the same generated password to hard-code into client.connect("arduinoClient", "username", "password") in mqtt_auth example, the connection will fail. I am still working on to automatically generate the password, not just hard-code, but I cannot understand why hard-code the password into the JAVA program works but fails in mquu_auth example.

With my Arduino UNO, I use [Cryptosuite](https://github.com/Cathedrow/Cryptosuite) library for hmac-sha1 and [arduino-base64](https://github.com/adamvr/arduino-base64) library for base64encode. Both of them are working fine separately. However, I need to get a base64-formatted output of hmac-sha1. So I have tried this:
```c
#include <sha1.h>
#include <Base64.h>

char key[] = "testKey";
char basestring[] = "testing";

void printHash(uint8_t* hash) {
int i;
for (i=0; i<20; i++) {
Serial.print("0123456789abcdef"[hash[i]>>4]);
Serial.print("0123456789abcdef"[hash[i]&0xf]);
}
Serial.println();
}

void setup() {
Serial.begin(115200);

Serial.print("Result:");
Sha1.initHmac((uint8_t*)key, strlen(key));
Sha1.print(basestring);
printHash(Sha1.resultHmac());
Serial.println();

// encoding
char* input;
input = (char*)(Sha1.resultHmac());
int inputLen = strlen(input);
int encodedLen = base64_enc_len(inputLen);
char encoded[encodedLen];

// note input is consumed in this step: it will be empty afterwards
base64_encode(encoded, input, inputLen); 

Serial.print("base64 result: ");
Serial.println(encoded);

}

void loop() {
}
```
The output of printHash that I got is 60d41271d43b875b791e2d54c34bf3f018a29763, which is exactly same with the online verification tool.
However, I supposed to get `YNQScdQ7h1t5Hi1Uw0vz8Biil2M=` for the base64 result. But I got `L18B0HicKRhuxmB6SIFpZP+DpHxU` which seems wrong. I have also tried this to the above JAVA program, which said that the output of the base64 result should be `YNQScdQ7h1t5Hi1Uw0vz8Biil2M=`

Basically my question is same as this post on Arduino forum: [Uint8_t to char array](http://forum.arduino.cc/index.php?topic=288726.0)
But seems like there's no further discussion about the solution, I repost my question there [http://forum.arduino.cc/index.php?topic=342528.0](http://forum.arduino.cc/index.php?topic=342528.0).

Don't know if there's any proper try I can do for this.
Any suggestion will be appreciated, thanks!

**Comment by knolleary** [comment]
So the problem is the base64 encoding of the password hash rather than with the mqtt client itself? That isn't something I've done, so I'm not going to be able to help debug that bit.

If you think there is an issue with the PubSubClient, I'd still like to see the actual Arduino code that involves the PubSubClient to see exactly how you're using it.

**Comment by e-lin** [comment]
Hi @knolleary 

A password is a string which contains about 330 characters and is generated by a process of OAuth encryption. Here is what I tried:
```c++
#include <SPI.h>
#include <Ethernet2.h>
#include <PubSubClient.h>

//MAC address of the Ethernet shield
byte mac[] = { 0x90, 0xA2, 0xD2, 0x0F, 0x97, 0xFF };
char macstr[] = "90a2d20f97ff";

//char servername[]= "test.mosquitto.org";
char servername[]= "cloudserver";
String clientName = String("d:quickstart:arduino:") + macstr;
char topicName[] = "hello/topic";
char username[] = "username";
char password[] = "QXV0aG9...";

EthernetClient ethClient;
PubSubClient client(servername, 1883, 0, ethClient);

void setup(){  
  Serial.begin(115200);
  if( 0 == Ethernet.begin(mac) ){
    Serial.println("Failed to configue Ethernet using DHCP");
  }
  
  //print the local IP address
  Serial.print("My IP address:");
  for( byte thisByte = 0; thisByte < 4; thisByte++ ){
    Serial.print(Ethernet.localIP()[thisByte], DEC);
    Serial.print(".");
  }
  Serial.println();
  Serial.println("Here");
}
  
void loop(){
  char clientStr[34];
  clientName.toCharArray(clientStr, 34);
  
  if (!client.connected()) {
    Serial.print("Trying to connect to: ");
    Serial.println(clientStr);
    Serial.print("Password: ");
    Serial.println(password);
    //client.connect(clientName);
    client.connect(clientStr, username, password);
  }

  if (client.connected() ) {
    char message[] = "Hi, there";
    
    //publish data
    boolean pubresult = client.publish(topicName,message);
    
    Serial.print("attempt to send ");
    Serial.println(message);
    Serial.print("to ");
    Serial.println(topicName);
    
    if (pubresult){
      Serial.println("successfully sent");
    }else{
      Serial.println("unsuccessfully sent");
    }
  }
  delay(5000);
}
```

And this is the output:
```
My IP address:192.128.1.163.
Here
Trying to connect to: d:quickstart:arduino:90a2d20f97ff
Password: 
QXV0aG9...
Trying to connect to: d:quickstart:arduino:90a2d20f97ff
Password: 
QXV0aG9...
Trying to connect to: d:quickstart:arduino:90a2d20f97ff
Password: 
QXV0aG9...
```

Without username and password, device can connect to test.mosquitto.org successfully. Hard-code the password "QXV0aG9..." into the testing JAVA program can connect to the cloud server successfully.

I am not sure if I had made a mistake?
Thank you.

**Comment by knolleary** [comment]
As documented, the client has a default maximum message size of 128 bytes - a password of 330 characters isn't going to fit in that.

Details on how to change this max size here: http://knolleary.net/arduino-client-for-mqtt/api/#configoptions

**Comment by e-lin** [comment]
Thanks.... check right now

**Comment by e-lin** [comment]
Originally I had the MQTT_MAX_PACKET_SIZE as 256, just set to 512 and had the Arduino IDE reboot.
Still got the same output as above..
512 should be enough.. the same password hard-code to java program can publish the message.

**Comment by knolleary** [comment]
At this point, I would use something like wireshark to compare the MQTT CONNECT packets being sent  by the arduino and the java client.

**Comment by e-lin** [comment]
Really thanks for suggestion. Later I'll try to use some tools to look into the MQTT CONNECT packets.

**Comment by e-lin** [comment]
After modified the maximum package size to fit the size of password, the mqtt auth example works. so closed this. Thanks knolleary for helping this.

**Comment by e-lin** [comment]
Hi @knolleary 
Maybe it's my misunderstanding that the mqtt_auth was working. Looking into the MQTT CONNECT packages, I found that the user name it got is strange. In the sketch I have the username like "magellanop.Project1". Somehow it just lost "ma" characters at the first two positions while calling client.connect(), and the payload of MQTT CONNECT package shows that it got "\357\277\275\002gellanop.Project1"

Any ideas?

**Comment by knolleary** [comment]
You don't explicitly null-terminate the client id, so it may be thinking it is longer than it really is.

    char clientStr[34];
    clientStr[33] = '\0';
    clientName.toCharArray(clientStr, 34);

**Comment by e-lin** [comment]
@knolleary 
No, I did not use char array. My application needs to publish data repeatedly, therefore I invoke a publish function in the loop() like below
```c
void loop(){
 if(!client.connected()){
    Serial.println(F("Reconnecting..."));
    reconnect();
  }else{
    Serial.println(F("Publish data..."));
    client.publish("outTopic","hello world");
  }
  client.loop();
}
```
Didn't look into very much maybe it's illegal?
The mqtt_auth example works fine when I only publish data once at the setup.

**Comment by knolleary** [comment]
@e-lin you asked about the connect packet looking invalid. From the example code you had provided earlier I provided what I thought was causing the connect packet to be invalid.

**Comment by e-lin** [comment]
@knolleary You're right, I'm sorry about that! I just had some tires, it is fine to publish data in loop(), the problem probably is `reconnect()` that I use a char pointer to hold the generated password..

```c
char* password;
void reconnect(){
  while(!client.connected()){
    if(client.connect("arduinoClient", "username", password )){
        client.publish("outTopic","hello world");
    }else{
      Serial.print(F("failed, rc="));
      Serial.print(client.state());
      Serial.println(F(" try again in 5 seconds"));
      delay(5000);
    }
  }
}
```

**Comment by knolleary** [comment]
I am even more confused now. You reopened this talking about a malformed username. Now you have hardcoded username that is different to what you mentioned before.

I don't know what code you are running and I don't know what issue you are hitting.

**Comment by e-lin** [comment]
Yes, have to close it.
